Hermes/appscale/hermes/stats/pubsub_base.py

```diff
@@ -0,0 +1,147 @@
+""" Base classes for building stats publishers and subscribers """
+import logging
+import sys
+
+from tornado.ioloop import PeriodicCallback, IOLoop
+
+
+class SubscriberIsAlreadyRegistered(Exception):
+  """ Raised when you register single subscriber twice on one publisher """
+  pass
+
+
+class SubscriberIsNotRegistered(Exception):
+  """ Raised when you unregister subscriber which is not registered yet """
+  pass
+
+
+class StatsSource(object):
+  """
+  Base class for producers of any kind of stats.
+  Subclasses should implement method get_current()
+  """
+  @property
+  def stats_name(self):
+    return self.__class__.__name__
+
+  def __repr__(self):
+    return self.stats_name
+
+  def get_current(self):
+    """ Returns: current value of specific kind of stats """
+    raise NotImplementedError()
+
+
+class AsyncStatsSource(object):
+  """
+  Base class for asynchronous producers of any kind of stats.
+  Subclasses should implement get_current_async()
+  """
+  @property
+  def stats_name(self):
+    return self.__class__.__name__
+
+  def __repr__(self):
+    return self.stats_name
+
+  def get_current_async(self):
+    """ Returns: Future wrapper for current value of specific kind of stats """
+    raise NotImplementedError()
+
+
+class StatsSubscriber(object):
+  """
+  Base class for stats consumers.
+  Subclasses should implement method receive(stats)
+  """
+  @property
+  def subscriber_name(self):
+      return self.__class__.__name__
+
+  def __repr__(self):
+    return self.subscriber_name
+
+  def receive(self, stats):
+    """ Handlers produced stats entity
+    Args:
+      stats: an object containing stats of some kind (node, processes, ...)
+    """
+    raise NotImplementedError()
+
+
+class StatsPublisher(object):
```

Comment:
In my opinion, the "Publish" and "Subscribe" terminology is a little bit misleading in this context. Most of the time I've seen these terms, they indicate that a subscriber is a client that maintains a persistent connection to a service that pushes content selectively. Since stats are obtained by clients using standard http requests, this might mislead a reader.

In this case, it appears that it is more like an abstraction over filtering what gets cached. Instead of "read and publish", it seems more like "collect and cache".